Added changes for supporting static imports

isaacrivriv opened the pull request:
fixes #22507

isaacrivriv committed:
Added changes for supporting static imports

isaacrivriv committed:
Merged Pages 31 import and directive interfaces into one and refactored code as needed for the merge

isaacrivriv committed:
Added import methods to tags generator

isaacrivriv committed:
Separated pages 3.1 API to its own bundle and added overlay for missing static imports

isaacrivriv commented:
#build

volosied commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/runtime/PagesDirectiveImportInfo.java:
```diff
@@ -14,10 +14,21 @@
+/**
+ * Interface for retrieving page directive information as well as page import information.
+ */
+public interface PagesDirectiveImportInfo {
```
```suggestion
public interface DirectiveInfo {
```

I think this should be a source of all directive information, so  I think it makes sense to remove `Import` from the name.

volosied replied:
Agh. Then again, these directives also apply to Tags. Maybe the name should just be DirectiveInfo?

volosied commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/GenerateTagFileVisitor.java:
```diff
@@ -222,7 +222,7 @@ protected void generateClassSection(ValidateTagFileResult validatorResult) {
 
         if(PagesVersionHandler.isPages31OrHigherLoaded()){
-            implementingInterfaces.add("com.ibm.ws.jsp.runtime.JspDirectiveInfo");
+            implementingInterfaces.add("com.ibm.ws.jsp.runtime.PagesDirectiveImportInfo");
```
```suggestion
            implementingInterfaces.add("com.ibm.ws.jsp.runtime.PagesDirectiveInfo");
```

volosied commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/GenerateJspVisitor.java:
```diff
@@ -244,8 +248,7 @@ protected void generateClassSection(ValidateJspResult validatorResult) {
         if(PagesVersionHandler.isPages31OrHigherLoaded()){
-            interfaces.add("com.ibm.ws.jsp.runtime.JspDirectiveInfo");
-            interfaces.add("com.ibm.ws.jsp.runtime.JspImportInfo");
+            interfaces.add("com.ibm.ws.jsp.runtime.PagesDirectiveImportInfo");
```
```suggestion
            interfaces.add("com.ibm.ws.jsp.runtime.PagesDirectiveInfo");
```

volosied commented on dev/io.openliberty.pages.3.1.internal.factories/src/org/apache/jasper/runtime/JspContextWrapper.java:
```diff
@@ -501,7 +501,7 @@ public ELContext getELContext() {
 
         // Pages 3.1
-        if(tag instanceof com.ibm.ws.jsp.runtime.JspDirectiveInfo && ((com.ibm.ws.jsp.runtime.JspDirectiveInfo)tag).isErrorOnELNotFound() ){
+        if(tag instanceof com.ibm.ws.jsp.runtime.PagesDirectiveImportInfo && ((com.ibm.ws.jsp.runtime.PagesDirectiveImportInfo)tag).isErrorOnELNotFound() ){
```
```suggestion
        if(tag instanceof com.ibm.ws.jsp.runtime.PagesDirectiveInfo && ((com.ibm.ws.jsp.runtime.PagesDirectiveInfo)tag).isErrorOnELNotFound() ){
```

volosied commented on dev/io.openliberty.pages.3.1.internal.factories/src/org/apache/jasper/runtime/PageContextImpl.java:
```diff
@@ -741,97 +762,103 @@ private static String XmlEscape(String s) {
-        if(servlet instanceof com.ibm.ws.jsp.runtime.JspImportInfo){
-            for(String _package : ((com.ibm.ws.jsp.runtime.JspImportInfo)servlet).getImportPackageList()){
+        if (servlet instanceof com.ibm.ws.jsp.runtime.PagesDirectiveImportInfo) {
+            for (String _package : ((com.ibm.ws.jsp.runtime.PagesDirectiveImportInfo) servlet).getImportPackageList()) {
+                System.out.println("Adding package: " + _package);
                 this.elContext.getImportHandler().importPackage(_package);
             }
-            for(String _class : ((com.ibm.ws.jsp.runtime.JspImportInfo)servlet).getImportClassList()){
+            for (String _class : ((com.ibm.ws.jsp.runtime.PagesDirectiveImportInfo) servlet).getImportClassList()) {
+                System.out.println("Adding class: " + _class);
                 this.elContext.getImportHandler().importClass(_class);
             }
             // not sure about importing static fields/methods? spec doesn't seem to mention it?
             // Follow up Liberty Issue 22507 to see if there's anything to be done here with static imports
+            // Does this work with importStatic com.example.*?
+            for (String _static : ((com.ibm.ws.jsp.runtime.PagesDirectiveImportInfo) servlet).getImportStaticList()) {
+                System.out.println("Adding static: " + _static);
+                this.elContext.getImportHandler().importStatic(_static);
+            }
         }
-
-		return this.elContext;
-	}
-    //LIDB4147-9 End
+    }
```
Looking over this again... Too much casting. haha   

Maybe just 
```
import com.ibm.ws.jsp.runtime.PagesDirectiveInfo;
...
PagesDirectiveInfo  pagesDirectiveInfo;
if (servlet instanceof PagesDirectiveInfo) {
     pagesDirectiveInfo  =  (PagesDirectiveInfo) servlet; 
}

from there just use pagesDirectiveInfo 
```

volosied reviewed (commented):
And don't for get to remove the System prints when this is ready :)

volosied commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/runtime/PagesDirectiveImportInfo.java:
```diff
@@ -14,10 +14,21 @@
+
+    // Import Info methods
 
     List<String> getImportClassList();
 
     List<String> getImportPackageList();
 
+    List<String> getImportStaticList();
```
I don't like having three methods for the one import, it would be nice to condense it into a single List (contains statics, package, and classes in separate groups).  Each directive would have a single method (but that's more future looking if there are new directives added) This can be done later on.  Something to think about later on.

isaacrivriv committed:
Fixed comments
- Renamed PageDirectiveImportInfo to PageDirectiveInfo
- Refactored code for changes regarding casting
- Removed prints
- Added check for 3.1 version

isaacrivriv committed:
Updated copyrights

isaacrivriv commented:
#build

isaacrivriv committed:
Added test for static imports and updated interface name to DirectiveInfo for shared pages and tags directives

volosied commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/ImportGenerator.java:
```diff
@@ -50,10 +50,13 @@ public void startGeneration(int section, JavaCodeWriter writer)
 					String directiveValue = attribute.getNodeValue();
 					if (directiveName.equals("import")) {
 						StringTokenizer tokenizer = new StringTokenizer(directiveValue, ",");
+						String STATIC_IMPORT_PREPPEND = "static ";
 						writeDebugStartBegin(writer);
 						while (tokenizer.hasMoreTokens()) {
 							String singleImport = ((String) tokenizer.nextToken()).trim();
-							if(singleImport.endsWith(".*")){
+							if (singleImport.startsWith(STATIC_IMPORT_PREPPEND)) {
+							        writer.println("importStaticList.add(\"" + singleImport.substring(STATIC_IMPORT_PREPPEND.length()) + "\");");
+							} else if(singleImport.endsWith(".*")){
```
If there's more than one space between static and the package, could that could break things? 

i.e `static   Math.PI`

isaacrivriv replied:
It did not, nice catch! Just pushed out a fix for it.

isaacrivriv committed:
Updated pages 3.1
- After quick fix in the jakarta api opensource for issue updated to pull in the latest 3.1.1 jar
- Removed proposed overlay
- Updated copyrights to latest usage

isaacrivriv committed:
Fix for static imports with trailing spaces

isaacrivriv commented:
#build

pnicolucci commented on dev/io.openliberty.jakarta/pages.3.1.bnd:
```diff
@@ -39,4 +36,4 @@ Import-Package: \
 
 -includeresource: \
-  @${repo;jakarta.servlet.jsp:jakarta.servlet.jsp-api;3.1.0;EXACT}!/!(META-INF/maven/*|module-info.class)
+  @${repo;jakarta.servlet.jsp:jakarta.servlet.jsp-api;3.1.1;EXACT}!/!(META-INF/maven/*|module-info.class)
```
new line

pnicolucci commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/GenerateTagFileVisitor.java:
```diff
@@ -250,7 +247,18 @@ protected void generateClassSection(ValidateTagFileResult validatorResult) {
+            // Cannot place this in the ImportGenerator since that is only run when the import directive is included in the page
+            // the imports below are required for all pages 
+            writer.println("static {");
+            // Since import directive is the same for tags as pages, need to verify that this is actually what we expect here
```
Can this comment be cleaned up now?

isaacrivriv committed:
Fixed comments

isaacrivriv commented:
#build

isaacrivriv commented:
#libby

isaacrivriv commented:
Build had only one error in `com.ibm.ws.classloading_test` unit tests which are unrelated to the changes. Should be safe to merge in

isaacrivriv commented:
Build had only one error in `com.ibm.ws.classloading_test` unit tests which are unrelated to the changes but it seems like it was affected by other issues and the FATs didn't run. Kicking off another build soon

isaacrivriv commented:
Re-requested build https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_lk50QKcREe24er7gxU0d3Q

volosied reviewed (approved):
Please merge!
